azure/services/async/async.go

```diff
@@ -174,6 +181,13 @@ func (s *Service) DeleteResource(ctx context.Context, spec azure.ResourceSpecGet
 	// No long running operation is active, so delete the resource.
 	log.V(2).Info("deleting resource", "service", serviceName, "resource", resourceName, "resourceGroup", rgName)
 	sdkFuture, err := s.Deleter.DeleteAsync(ctx, spec)
+
+	// If it is an intermittent failure with context deadline exceeded or canceled as the reconciler could not complete
+	// in the max amount of time, mark it as a transient error and return.
+	if err != nil && ContextDeadlineExceededOrCanceled(ctx) {
```

Comment:
Thinking a little bit more about this, do we want to handle `err != nil` and `ContextDeadlineExceededOrCanceled(ctx)` separately? 
Meaning, on an error, 
_if_ the `ContextDeadlineExceededOrCanceled(ctx)==true` and return with Transient error. 
_else_ you continue checking with `azure.ResourceNotFound(err)`. 

From current PR's line 183 till 207 could look like
```suggest
.
.
.
sdkFuture, err := s.Deleter.DeleteAsync(ctx, spec)
if err != nil { // check error first
  if azure.ResourceNotFound(err) {
    return nil // already deleted
  }

  if ContextDeadlineExceededOrCanceled(ctx) {
    return azure.WithTransientError(err, getRetryAfterFromError(err))
  }
  // none of the errors, bubble up this new error!
  return errors.Wrapf(err, "failed to delete resource %s/%s (service: %s)", rgName, resourceName, serviceName)
}

if sdkFuture != nil { // as its implemented already... 
  future, err := converters.SDKToFuture(sdkFuture, infrav1.DeleteFuture, serviceName, resourceName, rgName)
  if err != nil {
    return errors.Wrapf(err, "failed to delete resource %s/%s (service: %s)", rgName, resourceName, serviceName)
  }
  s.Scope.SetLongRunningOperationState(future)
  return azure.WithTransientError(azure.NewOperationNotDoneError(future), getRequeueAfterFromFuture(sdkFuture))
}

log.V(2).Info("successfully deleted resource", "service", serviceName, "resource", resourceName, "resourceGroup", rgName)
return nil
```

Same thoughts for the above. 
@CecileRobertMichon What do you think?

Reply:
holding off for this thought. Sorry for early lgtm :| 
/lgtm cancel

Reply:
hmm now that I'm also thinking about this more, doesn't this change short circuit `if sdkFuture != nil` ? We do a best effort `WaitForCompletion` then return the future + an error if the operation didn't complete within the timeout, for example: https://github.com/kubernetes-sigs/cluster-api-provider-azure/blob/main/azure/services/virtualmachines/client.go#L150

so this change would break our current async implementation because we would never store the future and would instead aggressively retry to create/delete, causing a higher number of API calls for each operation instead of doing the PUT/Create once and waiting for the operation to complete.

/hold

Reply:
So modifying @nawazkh's recommendation above, I think preserve the current order of first checking the sdkFuture, then processing remaining errors

```
	sdkFuture, err := s.Deleter.DeleteAsync(ctx, spec)
	if sdkFuture != nil {
		future, err := converters.SDKToFuture(sdkFuture, infrav1.DeleteFuture, serviceName, resourceName, rgName)
		if err != nil {
			return errors.Wrapf(err, "failed to delete resource %s/%s (service: %s)", rgName, resourceName, serviceName)
		}
		s.Scope.SetLongRunningOperationState(future)
		return azure.WithTransientError(azure.NewOperationNotDoneError(future), getRequeueAfterFromFuture(sdkFuture))
	} else if err != nil {
		if azure.ResourceNotFound(err) {
			// already deleted
			return nil
		}

 		if ContextDeadlineExceededOrCanceled(ctx) { // or maybe this becomes IsContextExceededError(err) ? 
		    return azure.WithTransientError(err, getRetryAfterFromError(err))
 		}

		return errors.Wrapf(err, "failed to delete resource %s/%s (service: %s)", rgName, resourceName, serviceName)
	}

	log.V(2).Info("successfully deleted resource", "service", serviceName, "resource", resourceName, "resourceGroup", rgName)
	return nil
```

Reply:
This is a good catch. Thank you.

Reply:
Thank you @CecileRobertMichon for adding context on `Futures`, current implementation with out this PR makes much more sense!